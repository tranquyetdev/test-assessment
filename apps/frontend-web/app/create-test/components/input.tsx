"use client"
import clsx from 'clsx';
import React from 'react'
import { Controller } from 'react-hook-form';

export type InputProps = {
  placeholder?: string;
  required?: boolean;
  label?: string;
  name: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  control: any;
  className?: string;
}

export default function Input({
  name,
  placeholder,
  required,
  label,
  control,
  className
}: InputProps) {

  return (
    <Controller
      name={name}
      control={control}
      rules={{ required: true }}
      render={({ field, formState: { errors } }) => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const errorField: any = errors;
        const errorMss = errorField[`${name}`]?.message;

        return (
          <div className={clsx("flex flex-col w-full", className)}>
            <label
              htmlFor={name}
              className="text-neutral-text-primary text-15 leading-24 mb-1"
            >{required && <span className="mr-1 text-error-base inline-block">*</span>}{label}</label>
            <input
              {...field}
              placeholder={placeholder}
              className={clsx("border border-solid py-[5px] px-3 outline-none text-15 leading-24 text-neutral-text-primary", {
                "border-error-base": errorMss,
                "border-neutral-border": !errorMss
              })}
            />
            {errorMss && <p className="mt-1 text-error-base">{errorMss}</p>}
          </div>
        );
      }}
    />
  )
}
