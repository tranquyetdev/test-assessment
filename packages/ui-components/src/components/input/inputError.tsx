import React from 'react';
import { Icon } from '../Icon';

export interface InputErrorProps {
  children: React.ReactNode;
}
export const InputError = ({ children }: InputErrorProps) => (
  <div className="flex space-x-1 items-center">
    <Icon name="caution" width={24} height={24} color="text-error" />
    <p className="text-error text-13 font-normal leading-6 font-primary">
      {children}
    </p>
  </div>
);
