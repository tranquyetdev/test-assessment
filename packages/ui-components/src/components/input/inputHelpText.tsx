import React from 'react';
import { Icon, IconName } from '../icons';

export interface InputHelpTextProps {
  children: React.ReactNode;
  variant: 'info' | 'error' | 'success';
}
export enum InputHelpTextVariant {
  INFO = 'info',
  ERROR = 'error',
  SUCCESS = 'success',
}

const variantToIconName: Record<string, string> = {
  [InputHelpTextVariant.ERROR]: 'caution',
  [InputHelpTextVariant.SUCCESS]: 'success',
};
const variantToTextColor: Record<string, string> = {
  [InputHelpTextVariant.ERROR]: 'text-error',
  [InputHelpTextVariant.SUCCESS]: 'text-success',
};

export const InputHelpText = ({ children, variant }: InputHelpTextProps) => {
  const iconName = variantToIconName[variant] as IconName;
  const textColor = variantToTextColor[variant];

  return (
    <div className="flex space-x-1 items-center">
      {iconName && <Icon name={iconName} style={{ color: textColor }} />}

      <p className={`${textColor} text-13 font-normal leading-6 font-primary`}>
        {children}
      </p>
    </div>
  );
};
