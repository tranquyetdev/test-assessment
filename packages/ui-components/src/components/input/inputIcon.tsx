import { Icon } from '../Icon';

export enum InputIconType {
  PASSWORD_HIDDEN = 'eyeShow',
  PASSWORD_VISIBLE = 'eyeHide',
  PROFILE = 'profile',
  CALENDAR = 'calendar',
}
export interface InputIconProps {
  type: string;
}

export const InputIcon = ({ type }: InputIconProps) => (
  <Icon name={type} width={24} height={24} color="text-neutral-placeholder" />
);
