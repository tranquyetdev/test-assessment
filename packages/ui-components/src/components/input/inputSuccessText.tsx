import React from 'react';
import { Icon } from '../Icon';

export interface InputSuccessProps {
  children: React.ReactNode;
}
export const InputSuccessText = ({ children }: InputSuccessProps) => (
  <div className="flex space-x-1 items-center h-7">
    <Icon name="success" width={24} height={24} color="text-success" />
    <p className="text-success text-13 font-normal leading-6 font-primary">
      {children}
    </p>
  </div>
);
