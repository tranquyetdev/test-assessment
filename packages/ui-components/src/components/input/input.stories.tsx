import type { Meta } from '@storybook/react';
import { Input as InputComponent } from './input';

const Story: Meta<typeof InputComponent> = {
  title: 'Components / Input',
  component: InputComponent,
};
export default Story;

export const Text = () => (
  <InputComponent
    placeholder="Please fill in your username"
    label="Username"
    width={400}
  />
);

export const Password = () => (
  <InputComponent
    label="Password"
    placeholder="Enter your password"
    type="password"
  />
);

export const Error = () => (
  <InputComponent
    error="this is sample error"
    value="this has error"
    label="Error"
  />
);
