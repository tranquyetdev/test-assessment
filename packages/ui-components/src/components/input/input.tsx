import React, { useState } from 'react';
import clsx from 'clsx';
import { InputHelpText } from './inputHelpText';
import { Icon, IconName } from '../icons';

export interface InputProps
  extends Omit<
    React.DetailedHTMLProps<
      React.InputHTMLAttributes<HTMLInputElement>,
      HTMLInputElement
    >,
    'size'
  > {
  label?: string;
  error?: string;
  fill?: boolean;
  leftIcon?: IconName;
  rightIcon?: IconName;
  size?: InputSize;
  infoText?: string;
  successText?: string;
  onChange: (value: React.ChangeEvent<HTMLInputElement>) => void;
  className?: string;
}
export enum InputType {
  PASSWORD = 'password',
  TEXT = 'text',
  SEARCH = 'search',
}
export enum InputSize {
  LARGE = 'lg',
  MEDIUM = 'md',
  SMALL = 'sm',
}

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (
    {
      label,
      error,
      type = InputType.TEXT,
      width = 400,
      fill,
      size = InputSize.MEDIUM,
      rightIcon,
      leftIcon,
      infoText,
      successText,
      value: rootValue,
      onChange,
      defaultValue = '',
      className,
      ...props
    }: InputProps,
    ref
  ) => {
    const [currentType, setCurrentType] = useState(type);
    const [value, setValue] = useState(defaultValue);
    const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      if (rootValue === undefined) {
        return setValue(event?.target?.value);
      }
      onChange && onChange(event);
    };

    const isHelpTextVisible = error || infoText || successText;

    return (
      <div className="space-y-2">
        {label && (
          <label
            htmlFor={props.name}
            className="block text-13 leading-6 font-medium font-primary text-neutral-placeholder"
          >
            {label}
            {props.required && (
              <span className="pl-[2px] text-error-base">*</span>
            )}
          </label>
        )}

        <div
          className={clsx(
            'flex border-[1px]',
            'text-13 leading-6 font-normal font-primary',
            'placeholder-neutral-border',
            'hover:border-neutral-placeholder',
            'focus-within:border-primary',
            'hover:focus-within:border-primary',
            !rootValue && 'border-neutral-border',
            error && '!border-error-border bg-error-bg',
            successText && 'border-success-border bg-success-bg',
            type === InputType.SEARCH
              ? 'px-0 text-neutral-white bg-neutral-text-secondary !border-0'
              : 'px-3',
            props.disabled &&
            '!border-none !bg-neutral-disable !text-neutral-placeholder',
            className
          )}
          style={{ width: fill ? '100%' : width }}
        >
          {leftIcon && (
            <div className="flex items-center text-neutral-text-secondary pr-3">
              <Icon name={leftIcon} />
            </div>
          )}

          <input
            {...props}
            value={rootValue === undefined ? value : rootValue}
            onChange={handleOnChange}
            ref={ref}
            id={props.name}
            type={currentType}
            className={clsx(
              'focus:outline-none bg-transparent w-full',
              {
                [InputSize.LARGE]: 'py-3',
                [InputSize.MEDIUM]: 'py-2',
                [InputSize.SMALL]: 'py-1',
              }[size],
              // TODO: apply background opacity
              type === InputType.SEARCH &&
                'px-3 focus:border-neutral-white focus:border bg-neutral-text-primary',
              className
            )}
            aria-invalid={Boolean(error)}
            aria-describedby={error ? `${props.name}-error` : undefined}
          />

          {type === InputType.PASSWORD && (
            <button
              className="flex items-center pl-2 text-neutral-text-secondary"
              onClick={() => {
                setCurrentType((type) =>
                  type === 'password' ? InputType.TEXT : InputType.PASSWORD
                );
              }}
            >
              <Icon
                name={
                  currentType === InputType.PASSWORD ? 'eye-show' : 'eye-hide'
                }
              />
            </button>
          )}

          {type === InputType.SEARCH && (
            <button
              className={clsx(
                'flex items-center bg-primary-base text-neutral-text-secondary',
                {
                  [InputSize.LARGE]: 'px-3',
                  [InputSize.MEDIUM]: 'px-2',
                  [InputSize.SMALL]: 'px-1',
                }[size]
              )}
              // @ts-expect-error todo
              onClick={props.onSubmit}
            >
              <Icon name="search" strokeWidth={40} />
            </button>
          )}

          {rightIcon && (
            <div className="flex items-center text-neutral-text-secondary pl-2">
              <Icon name={rightIcon} />
            </div>
          )}
        </div>

        {(error || infoText || successText) && <div className={isHelpTextVisible ? 'opacity-100' : 'opacity-0'}>
          {error && <InputHelpText variant="error">{error}</InputHelpText>}
          {infoText && <InputHelpText variant="info">{infoText}</InputHelpText>}
          {successText && (
            <InputHelpText variant="success">{successText}</InputHelpText>
          )}
        </div>}
      </div>
    );
  }
);
