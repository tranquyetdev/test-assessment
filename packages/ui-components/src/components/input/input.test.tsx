import { render } from '@testing-library/react';

import { Input } from './input';

describe('Button', () => {
  it(`should render input successfully`, () => {
    const { baseElement } = render(
      <Input type="text" value="this is value" width={300} />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
