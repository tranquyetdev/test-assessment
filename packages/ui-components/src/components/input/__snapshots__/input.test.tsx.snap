// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render input successfully 1`] = `
<body>
  <div>
    <div
      class="space-y-2"
    >
      <div
        class="flex border-[1px] text-13 leading-6 font-normal font-primary placeholder-neutral-border hover:border-neutral-placeholder focus-within:border-primary hover:focus-within:border-primary border-neutral-border px-3"
        style="width: 300px;"
      >
        <input
          aria-invalid="false"
          class="focus:outline-none bg-transparent w-full py-2"
          type="text"
          value="this is value"
        />
      </div>
      <div
        class="opacity-0"
      />
    </div>
  </div>
</body>
`;
